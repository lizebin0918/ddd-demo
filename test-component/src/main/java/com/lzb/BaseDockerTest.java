package com.lzb;

import java.time.Duration;
import java.time.Instant;
import java.util.stream.Stream;

import com.lzb.initializer.DataSourceInitializer;
import com.lzb.initializer.RedisInitializer;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.TestInfo;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.junit.jupiter.Testcontainers;

import org.springframework.test.context.ContextConfiguration;

import static com.lzb.initializer.DataSourceInitializer.database;
import static com.lzb.initializer.RedisInitializer.redis;

@Slf4j
@Testcontainers(parallel = true)
@ContextConfiguration(initializers = {
        DataSourceInitializer.class,
        RedisInitializer.class
})
public abstract class BaseDockerTest extends BaseTest {

    static {

        Instant start = Instant.now();
        Stream.of(database, redis).parallel().forEach(GenericContainer::start);
        log.info("----------------------------------------------------------------------");
        log.info("TestContainers 启动耗时 {} 秒", Duration.between(start, Instant.now()).getSeconds());
        log.info("----------------------------------------------------------------------");

    }

    @AfterAll
    public static void afterAll(TestInfo testInfo) {
    }

    /**
     * 每个test类都会执行一次
     */
    @BeforeAll
    public static void beforeAll(TestInfo testInfo) {
    }

    /**
     * 每个方法执行之前执行一次
     */
    @BeforeEach
    public void beforeEach() {
    }

}