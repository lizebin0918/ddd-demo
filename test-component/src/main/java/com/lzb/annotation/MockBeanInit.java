package com.lzb.annotation;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;


/**
 * 用于给使用@MockBean或者@SpyBean注解的bean自定义mock逻辑,防止在mock逻辑执行之前就被调用
 * Notice:只有在第一个Test方法执行生效，第一个测试方法执行完之后会被reset
 * 1. 只能加到静态方法上
 * 2. 方法入参只能1个
 * 3. 方法入参必须为@MockBean或者@SpyBean注解的类型
 * @author lizebin
 * @date 2023/07/22
 */
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface MockBeanInit {

}
