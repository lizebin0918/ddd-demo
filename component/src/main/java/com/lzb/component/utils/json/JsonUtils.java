package com.lzb.component.utils.json;

import java.math.BigDecimal;
import java.time.ZoneId;
import java.util.Locale;
import java.util.Objects;
import java.util.TimeZone;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.json.JsonReadFeature;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.databind.ser.std.ToStringSerializer;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import io.vavr.jackson.datatype.VavrModule;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.Nullable;

@UtilityClass
@Slf4j
public class JsonUtils {

    public static final JsonMapper OBJECT_MAPPER = JsonMapper.builder()
            .configure(JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS, true)
            .configure(JsonReadFeature.ALLOW_BACKSLASH_ESCAPING_ANY_CHARACTER, true)
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
            .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
            // 枚举写成字符串
            .configure(SerializationFeature.WRITE_ENUMS_USING_TO_STRING, true)
            // 时间序列化
            .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
            // 时区
            .defaultLocale(Locale.CHINA)
            .defaultTimeZone(TimeZone.getTimeZone(ZoneId.systemDefault()))
            .addModule(new JavaTimeModule())
            .addModule(new Jdk8Module())
            // 通过构造函数声明@JsonCreator反序列化，但是要在maven引入compile插件，加参数:-parameters
            .addModule(new ParameterNamesModule(JsonCreator.Mode.PROPERTIES))
            .addModule(new SimpleModule().addSerializer(BigDecimal.class, new ToStringSerializer()))
            .addModule(new VavrModule())
            .build();

   static {
        // 输入非空字段
        OBJECT_MAPPER.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        // 只显示有的字段
        OBJECT_MAPPER.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE);
        OBJECT_MAPPER.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
    }

    /**
     * 将对象序列化成json字符串
     *
     * @param value javaBean
     * @param <T>   T 泛型标记
     * @return jsonString json字符串
     */
    @Nullable
    public static <T> String toJSONString(T value) {
        try {
            return OBJECT_MAPPER.writeValueAsString(value);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }



    /**
     * 将json反序列化成对象
     *
     * @param content   content
     * @param valueType class
     * @param <T>       T 泛型标记
     * @return Bean
     */
    public static <T> T json2JavaBean(String content, Class<T> valueType) {
        if (Objects.isNull(content)) {
            return null;
        }
        try {
            return OBJECT_MAPPER.readValue(content, valueType);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            throw new RuntimeException(e);
        }
    }

}