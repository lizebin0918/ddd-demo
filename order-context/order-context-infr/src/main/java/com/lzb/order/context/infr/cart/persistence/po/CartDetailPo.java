package com.lzb.order.context.infr.cart.persistence.po;

import java.io.Serializable;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.lzb.component.mybatis.BasePo;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper = true)
@TableName("cart_detail")
public class CartDetailPo extends BasePo implements Serializable {

    public static final long DEFAULT_DELETE_TIMESTAMP = 0L;

    @TableId(value = "id", type = IdType.AUTO)
    private Long id;

    private Long userId;

    private Integer skuId;

    /**
     * 数量
     */
    private Integer count;

    /**
     * 删除时间戳
     */
    private Long deleteTimestamp;

    /**
     * 删除标识
     */
    private Boolean deleteFlag;


    @Override
    public Serializable id() {
        return this.id;
    }
}