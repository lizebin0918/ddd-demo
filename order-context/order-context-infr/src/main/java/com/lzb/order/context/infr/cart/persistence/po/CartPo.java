package com.lzb.order.context.infr.cart.persistence.po;

import java.io.Serializable;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.annotation.Version;
import com.lzb.component.mybatis.BasePo;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper = true)
@TableName("cart")
@AllArgsConstructor
public class CartPo extends BasePo implements Serializable {

    public static final int DEFAULT_DELETE_VERSION = 1;

    @TableId(type = IdType.AUTO)
    private Long userId;

    @Version
    private Integer version;

    @Override
    public Serializable id() {
        return userId;
    }
}