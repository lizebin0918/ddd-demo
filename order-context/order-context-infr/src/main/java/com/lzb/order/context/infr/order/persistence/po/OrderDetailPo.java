package com.lzb.order.context.infr.order.persistence.po;

import java.io.Serializable;
import java.math.BigDecimal;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.lzb.component.mybatis.BasePo;
import com.lzb.order.context.domain.order.aggregation.valobj.OrderStatus;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper = false)
@TableName("order_detail")
public class OrderDetailPo extends BasePo {

    @TableId(type = IdType.AUTO)
    private Long id;

    private Long orderId;

    private Integer skuId;

    private OrderStatus orderStatus;

    private BigDecimal price;

    private Boolean locked;


    @Override
    public Serializable id() {
        return id;
    }
}
