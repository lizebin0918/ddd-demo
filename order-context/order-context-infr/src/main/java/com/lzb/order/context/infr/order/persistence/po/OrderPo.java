package com.lzb.order.context.infr.order.persistence.po;

import java.io.Serializable;
import java.math.BigDecimal;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.annotation.Version;
import com.lzb.component.mybatis.BasePo;
import com.lzb.component.mybatis.handler.String2JsonTypeHandler;
import com.lzb.order.context.domain.order.aggregation.valobj.OrderStatus;
import lombok.Data;
import lombok.EqualsAndHashCode;

/**
 * <p>
 * 
 * </p>
 *
 * @author lizebin
 * @since 2023-08-29
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("\"order\"")
public class OrderPo extends BasePo {

    @TableId(type = IdType.INPUT)
    private Long orderId;

    private OrderStatus orderStatus;

    private String currency;

    /**
     * 汇率
     */
    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    @Version
    private int version;

    ///////////////////////////////////////////////////////////////////////////
    // 地址相关
    ///////////////////////////////////////////////////////////////////////////

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    private String firstName;

    private String lastName;

    /**
     * 地址1
     */
    private String addressLine1;

    /**
     * 地址2
     */
    private String addressLine2;

    /**
     * 国家
     */
    private String country;

    /**
     * 扩展
     */
    @TableField(typeHandler = String2JsonTypeHandler.class)
    private String extension;

    @Override
    public Serializable id() {
        return orderId;
    }
}