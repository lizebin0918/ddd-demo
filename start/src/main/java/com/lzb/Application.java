package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.autoconfigure.data.redis.RedisRepositoriesAutoConfiguration;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.scheduling.annotation.EnableAsync;

// 启动报错：Multiple Spring Data modules found, entering strict repository configuration
@SpringBootApplication(exclude = { RedisRepositoriesAutoConfiguration.class})
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
