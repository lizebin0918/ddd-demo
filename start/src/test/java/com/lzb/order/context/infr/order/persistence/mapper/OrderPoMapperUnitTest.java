package com.lzb.order.context.infr.order.persistence.mapper;

import java.util.List;
import java.util.Map;

import com.baomidou.mybatisplus.core.toolkit.Wrappers;
import com.lzb.BaseMapperUnitTest;
import com.lzb.component.utils.json.JsonUtils;
import com.lzb.order.context.infr.order.persistence.po.OrderPo;
import org.jeasy.random.EasyRandom;
import org.jeasy.random.EasyRandomParameters;
import org.json.JSONException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.skyscreamer.jsonassert.JSONAssert;

import org.springframework.beans.factory.annotation.Autowired;

/**
 * <br/>
 * Created on : 2023-09-18 22:31
 * @author lizebin
 */
class OrderPoMapperUnitTest extends BaseMapperUnitTest {

    @Autowired
    private OrderPoMapper orderPoMapper;

    @Test
    @DisplayName("查询订单表")
    void should_count() {
        assertDoesNotThrow(() -> {
            orderPoMapper.selectCount(Wrappers.emptyWrapper());
        });
    }

    @Test
    @DisplayName("测试直接写xml的sql，看看extension查询是否有问题")
    void should_query_sql_in_mapper() throws JSONException {

        EasyRandomParameters parameters = new EasyRandomParameters();
        EasyRandom generator = new EasyRandom(parameters);
        OrderPo orderPo = generator.nextObject(OrderPo.class);
        orderPo.setExtension(JsonUtils.toJSONString(Map.of("name", "hippo")));


        System.out.println(JsonUtils.toJSONString(orderPo));

        orderPoMapper.insert(orderPo);
        long orderId = orderPo.getOrderId();
        OrderPo actualOrderPo = orderPoMapper.get(orderId);

        JSONAssert.assertEquals(actualOrderPo.getExtension(), orderPo.getExtension(), false);
    }

    @Test
    @DisplayName("测试返回值类型为map的查询")
    void should_order_count() {
        EasyRandom generator = new EasyRandom();
        OrderPo orderPo = generator.nextObject(OrderPo.class);
        orderPo.setExtension(null);

        orderPoMapper.insert(orderPo);
        List<Map<String, Integer>> o = orderPoMapper.orderCount();
        System.out.println(o.getFirst().size());
        System.out.println(o.getFirst().get(1L));
        assertJSON(o);
    }

    @Test
    @DisplayName("直接查询string")
    void should_query_string() {
        assertThat(orderPoMapper.selectStr()).isEqualTo("1");
    }

}
