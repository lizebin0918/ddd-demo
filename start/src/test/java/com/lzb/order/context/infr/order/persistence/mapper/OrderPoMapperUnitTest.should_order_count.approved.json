[
  {
    "count": 1,
    "orderid": -5106534569952410475
  }
]